benchmark/spmv/cuda_linops.hpp

```diff
@@ -0,0 +1,468 @@
+    {
+#ifdef ALLOWMP
+        algmode_ = CUSPARSE_ALG_MERGE_PATH;
+#endif
+    }
+
+    ~CuspCsrEx() override
+    {
+        const auto id = this->get_gpu_exec()->get_device_id();
+        gko::device_guard g{id};
+        if (set_buffer_) {
+            try {
+                GKO_ASSERT_NO_CUDA_ERRORS(cudaFree(buffer_));
+            } catch (std::exception &e) {
+                std::cout
+                    << "Error when unallocating CuspCsrEx temporary buffer"
+                    << std::endl;
+            }
+        }
+    }
+
+private:
+    // Contains {alpha, beta}
+    gko::Array<ValueType> scalars{
+        this->get_executor(), {gko::one<ValueType>(), gko::zero<ValueType>()}};
+    std::shared_ptr<csr> csr_;
+    cusparseOperation_t trans_;
+    cusparseAlgMode_t algmode_;
+    void *buffer_;
+    bool set_buffer_;
+};
+
+
+template <typename ValueType = gko::default_precision,
+          typename IndexType = gko::int32,
+          cusparseHybPartition_t Partition = CUSPARSE_HYB_PARTITION_AUTO,
+          int Threshold = 0>
+class CuspHybrid
+    : public gko::EnableLinOp<
+          CuspHybrid<ValueType, IndexType, Partition, Threshold>, CuspBase>,
+      public gko::EnableCreateMethod<
+          CuspHybrid<ValueType, IndexType, Partition, Threshold>>,
+      public gko::ReadableFromMatrixData<ValueType, IndexType> {
+    friend class gko::EnableCreateMethod<CuspHybrid>;
+
+public:
+    using csr = gko::matrix::Csr<ValueType, IndexType>;
+    using mat_data = gko::matrix_data<ValueType, IndexType>;
+
+    void read(const mat_data &data) override
+    {
+        auto t_csr = csr::create(this->get_executor(),
+                                 std::make_shared<typename csr::classical>());
+        t_csr->read(data);
+        this->set_size(gko::dim<2>{t_csr->get_size()});
+
+        const auto id = this->get_gpu_exec()->get_device_id();
+        gko::device_guard g{id};
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(cusparseDcsr2hyb(
+            this->get_gpu_exec()->get_cusparse_handle(), this->get_size()[0],
+            this->get_size()[1], this->get_descr(), t_csr->get_const_values(),
+            t_csr->get_const_row_ptrs(), t_csr->get_const_col_idxs(), hyb_,
+            Threshold, Partition));
+    }
+
+    void apply_impl(const gko::LinOp *b, gko::LinOp *x) const override
+    {
+        auto dense_b = gko::as<gko::matrix::Dense<ValueType>>(b);
+        auto dense_x = gko::as<gko::matrix::Dense<ValueType>>(x);
+        auto db = dense_b->get_const_values();
+        auto dx = dense_x->get_values();
+
+        const auto id = this->get_gpu_exec()->get_device_id();
+        gko::device_guard g{id};
+        GKO_ASSERT_NO_CUSPARSE_ERRORS(
+            cusparseDhybmv(this->get_gpu_exec()->get_cusparse_handle(), trans_,
+                           &scalars.get_const_data()[0], this->get_descr(),
+                           hyb_, db, &scalars.get_const_data()[1], dx));
+    }
+
+    CuspHybrid(std::shared_ptr<const gko::Executor> exec,
```

Comment:
Also here: Can you make this `protected`, so the user must use the `::create()` method?